### ISSUE TITLE
Support `undici` ProxyAgent

### ISSUE THREAD

**Issue by Hebilicious** [oven-sh/bun#4474]
## Workaround

There's a fetch workaround available.

<img width="415" alt="image" src="https://github.com/oven-sh/bun/assets/13395944/ecfc6ef9-7a6c-4d31-ac78-0a771fcbc86d">

### What version of Bun is running?

0.8.1+16b4bf341acc0f4804f0b6bdf5298c180cd00366

### What platform is your computer?

Darwin 22.6.0 arm64 arm

### What steps can reproduce the bug?

The following code works with node, but doesn't work with bun : 

Undici ProxyAgent : 

```ts
import { ProxyAgent, type RequestInit as UndiciRequestInit, fetch as uFetch } from "undici"

export const fetchWithProxyUndici = async (url: string, options: UndiciRequestInit, proxy: URL) => {
  const dispatcher = new ProxyAgent({
    uri: proxy.host,
    token: `Basic ${Buffer.from(`${proxy.username}:${proxy.password}`).toString("base64")}`
  })
  return uFetch(url, { ...options, dispatcher })
}
```

Something like this should work but it doesn't  : 

```ts
import http from "node:http"
import https from "node:https"
export async function fetchWithNodeProxy(url: string, options: Record<string, any>, proxy: URL) {
  return new Promise<{ text: () => Promise<string>; headers: Headers; status?: number; ok: boolean }>((resolve, reject) => {
    const urlParsed = new URL(url)
    const headers = options.headers || {}
    headers["Proxy-Authorization"] = `Basic ${Buffer.from(`${proxy.username}:${proxy.password}`).toString("base64")}`
    const requestOptions: http.RequestOptions = {
      hostname: proxy.hostname,
      port: proxy.port,
      method: "CONNECT",
      path: `${urlParsed.host}:443`,
      headers
    }
    http.request(requestOptions)
      .on("connect", (res, socket) => {
        console.log("RES !", res.statusCode)
        delete headers["Proxy-Authorization"]
        if (res.statusCode === 200) {
          const agent = new https.Agent({ socket })
          const req = https.get({
            host: urlParsed.hostname,
            path: urlParsed.pathname,
            agent,
          }, (response) => {
            const chunks = [] as Uint8Array[]
            response.on("data", (chunk) => {
              chunks.push(chunk)
            })
            response.on("end", () => {
              console.log("Succesful response !")
              resolve({
                ok: true,
                headers: new Headers(response.headers as any),
                status: response.statusCode,
                text: async () => Buffer.concat(chunks).toString()
              })
            })

            response.on("error", (err) => {
              console.log("Error response !")
              reject(err)
            })

            response.setTimeout(15000, () => {
              reject(new Error("Request timed out"))
            })
          })

          req.on("error", (err) => {
            console.log("Error request !")
            reject(err.message)
          })
        }
        else {
          reject(new Error("Could not connect to proxy!"))
        }
      }).on("error", (err) => {
        console.log("Error connect !")
        reject(err.message)
      }).end()
  })
}
}
```

### What is the expected behavior?

With Node v20.5 everything works as expected and the Request is sent through the proxy.

### What do you see instead?

With Bun I get : 

For undici : 
```
SyntaxError: Import named 'ProxyAgent' not found in module 'undici'.
```
For http + connect
```
Error connect !
error: fetch() URL is invalid
```

**Comment by Hanaasagi** [comment]
I fixed a similar issue last week(https://github.com/oven-sh/bun/issues/4295). It seems that it doesn't cover all cases. node supports that path is a url without a protocol prefix. 🥲

```
      path: `${urlParsed.host}:443`
```

----


Another issue is that the `CONNECT` event is not triggered, even without proxy.

```JavaScript
const http = require("http");

const options = {
  hostname: "example.com",
  port: 80,
  path: "/",
  method: "CONNECT",
};

const req = http.request(options);

req.on("connect", (res, socket, head) => {
  console.log("Connect!");
});

req.end();

```

node outputs "Connect!", but Bun 0.8.1 does not.

**Comment by Hebilicious** [comment]
> I fixed a similar issue last week(#4295). It seems that it doesn't cover all cases. node supports that path is a url without a protocol prefix. 🥲
> 
> ```
>       path: `${urlParsed.host}:443`
> ```
> 
> Another issue is that the `CONNECT` event is not triggered, even without proxy.
> 
> ```js
> const http = require("http");
> 
> const options = {
>   hostname: "example.com",
>   port: 80,
>   path: "/",
>   method: "CONNECT",
> };
> 
> const req = http.request(options);
> 
> req.on("connect", (res, socket, head) => {
>   console.log("Connect!");
> });
> 
> req.end();
> ```
> 
> node outputs "Connect!", but Bun 0.8.1 does not.

Thanks for the insight. I think the "easiest" way to unlock proxy usage would be to support undici dispatcher, as it's currently the most performant and simple option for node.js.

**Comment by Hanaasagi** [comment]
I noticed that undici is rewrited in Bun and there are some APIs that have not been implemented. Related to this issue https://github.com/oven-sh/bun/issues/1987

Source File: https://github.com/oven-sh/bun/blob/main/src/js/thirdparty/undici.js

**Comment by Hebilicious** [comment]
> I noticed that undici is rewrited in Bun and there are some APIs that have not been implemented. Related to this issue #1987
> 
> Source File: [`main`/src/js/thirdparty/undici.js](https://github.com/oven-sh/bun/blob/main/src/js/thirdparty/undici.js?rgh-link-date=2023-09-04T15%3A02%3A35Z)

Does that mean that ProxyAgent/dispatcher just needs to be implemented on the JS side ? (ie no c++ /zig stuff?) If so I'm happy to contribute that.

**Comment by Hanaasagi** [comment]
I'm not sure why undici has been placed in the thirdparty. This might require someone from the core team to explain. Perhaps you can ask on Discord for more information.

**Comment by Hebilicious** [comment]
> I'm not sure why undici has been placed in the thirdparty. This might require someone from the core team to explain. Perhaps you can ask on Discord for more information.

Looking at the discord it looks like we can do this though : 

<img width="415" alt="image" src="https://github.com/oven-sh/bun/assets/13395944/ecfc6ef9-7a6c-4d31-ac78-0a771fcbc86d">

**Comment by Hanaasagi** [comment]
Latest development guide: https://github.com/oven-sh/bun/blob/main/docs/project/development.md

The first time, compile the whole project with `make setup`. After the first compilation, if you only modify the https://github.com/oven-sh/bun/blob/main/src/js/thirdparty/undici.js file, just run `make js`, you can see the change.

**Comment by ThatOneBro** [comment]
> I'm not sure why undici has been placed in the thirdparty. 

Basically the `undici` implementation is a wrapper around the Bun implementation of fetch. We did that so that `undici` wouldn't be slow basically, since the Bun implementation of `fetch` is super fast.

Sadly there are some concepts in `undici` which don't translate well, but there's definitely some hacks that can make it work.

I think the proxy thing shouldn't be too hard to get working if it already works in `fetch`. I think the idea would be to create a fake `ProxyAgent` and just use it as a context wrapper that keeps the props related to proxy config in there and when dispatcher is a `ProxyAgent` just use the `proxy` prop in the `fetch` options.

**Comment by faintsmi1e** [comment]
> > I'm not sure why undici has been placed in the thirdparty.
> 
> Basically the `undici` implementation is a wrapper around the Bun implementation of fetch. We did that so that `undici` wouldn't be slow basically, since the Bun implementation of `fetch` is super fast.
> 
> Sadly there are some concepts in `undici` which don't translate well, but there's definitely some hacks that can make it work.
> 
> I think the proxy thing shouldn't be too hard to get working if it already works in `fetch`. I think the idea would be to create a fake `ProxyAgent` and just use it as a context wrapper that keeps the props related to proxy config in there and when dispatcher is a `ProxyAgent` just use the `proxy` prop in the `fetch` options.

does  `proxy` prop in the `fetch` options supports socks5 proxy? It looks like it's not working

**Comment by akshitdandyan** [comment]
Any updates over this issue? Proxy is not working for me, tried fetch, axios, gaxios & https-proxy-agent.

**Comment by gnosticdev** [comment]
Can now use Bun Shell for this also. 

This is running through my vpn (mullvad) via [gluetun](https://github.com/qdm12/gluetun) as the proxy server.

```ts
import { $ } from 'bun'

function proxyRequest(requestURL: string) {
    const url = new URL('http://127.0.0.1:8888')
    url.username = 'user'
    url.password = 'password'

    return $`curl --proxy ${url.toString()} ${requestURL}`.json()
}

const out = await proxyRequest('https://ipinfo.io')
```

**Comment by javeoff** [comment]
upvote

**Comment by Niklas2290** [comment]
+ 1

**Comment by nospher3x2** [comment]
upvote too

**Comment by Jarred-Sumner** [comment]
Updated this issue to reflect that `ProxyAgent` with `undici` is not supported yet.

To send an HTTP request using a proxy in Bun, you can:
- Use `node:http`
- [Pass `proxy` to `fetch()`](https://bun.sh/guides/http/proxy)

**Comment by nospher3x2** [comment]
> Updated this issue to reflect that `ProxyAgent` with `undici` is not supported yet.
> 
> To send an HTTP request using a proxy in Bun, you can:
> 
> * Use `node:http`
> * [Pass `proxy` to `fetch()`](https://bun.sh/guides/http/proxy)

but has other problem that `proxy` in`fetch` doesn't work on windows :c

**Comment by Zamion101** [comment]
> Can now use Bun Shell for this also.
> 
> This is running through my vpn (mullvad) via [gluetun](https://github.com/qdm12/gluetun) as the proxy server.
> 
> ```ts
> import { $ } from 'bun'
> 
> function proxyRequest(requestURL: string) {
>     const url = new URL('http://127.0.0.1:8888')
>     url.username = 'user'
>     url.password = 'password'
> 
>     return $`curl --proxy ${url.toString()} ${requestURL}`.json()
> }
> 
> const out = await proxyRequest('https://ipinfo.io')
> ```

I also tried using this solution. But curl refuses to utilize `--proxy` argument when it's called from `$`. But when I try to run command in a shell it correctly uses socks5 proxy. Is it a bug in bun?

Here is my curl command `curl  --http1.1 --proxy socks5://username:password@XXX.XXX.XXX.XXX:1080 https://httpbin.org/ip`. it should return `XXX.XXX.XXX.XXX` in response but it returns my current ip address. I also tried to see what happens when called using `-v` but curl looks like doesn't even recognizes there is a --proxy argument given.

`When runned from $`
![image](https://github.com/oven-sh/bun/assets/8071263/14146eda-8707-4a49-b902-78cd4f6ce6d4)

`Same command when run in bash`
![image](https://github.com/oven-sh/bun/assets/8071263/ee6d3b5d-b7d2-4edb-9ca4-6f4e092eee5a)

**Comment by rhuanbarreto** [comment]
Package upgrade @octokit/action to v7 is blocked by this.

**Comment by catFurr** [comment]
I ran into this issue recently trying to use Appium with the webdriverio driver with Bun v1.1.34

Does anyone have any idea if/when this can be resolved? Thanks!


For reference, I used this script:
Make sure to replace `deviceName` and `udid` with proper values.
```js
import { remote } from 'webdriverio';

async function testAppiumConnection() {
  const capabilities = {
    platformName: 'Android',
    'appium:automationName': 'UiAutomator2',
    'appium:deviceName': 'REDACTED',
    'appium:udid': 'REDACTED',
    'appium:noReset': true,
  };

  try {
    console.log('Attempting to connect to device...');
    const driver = await remote({
      protocol: 'http',
      hostname: '127.0.0.1',
      port: 4723,
      path: '/',
      capabilities: capabilities,
      logLevel: 'info',
    });

    // If we get here, connection was successful
    console.log('Successfully connected to device!');

    // Get some basic device info to verify
    const deviceInfo = await driver.getDeviceTime();
    console.log('Device time:', deviceInfo);

    // Clean up
    await driver.deleteSession();
    console.log('Session ended successfully');
  } catch (error) {
    console.error('Error connecting to device:', error.message);
  }
}

testAppiumConnection();
```

Result:
```
1 | (function (entry, fetcher)
    ^
SyntaxError: Export named 'ProxyAgent' not found in module 'undici'.
```

**Comment by jlucaso1** [comment]
@catFurr can you try on latest version? (1.2.5). I tested running  your code and got the same output as node:

```bash
02:44:49 jlucaso@jlucaso bun-3997 → node index.ts 
(node:325521) ExperimentalWarning: Type Stripping is an experimental feature and might change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
Attempting to connect to device...
2025-03-14T05:45:11.174Z INFO webdriver: Initiate new session using the WebDriver protocol
2025-03-14T05:45:11.175Z INFO @wdio/utils: Connecting to existing driver at http://127.0.0.1:4723/
2025-03-14T05:45:11.181Z INFO webdriver: [POST] http://127.0.0.1:4723/session
2025-03-14T05:45:11.181Z INFO webdriver: DATA {
  capabilities: {
    alwaysMatch: {
      platformName: 'Android',
      'appium:automationName': 'UiAutomator2',
      'appium:deviceName': 'REDACTED',
      'appium:udid': 'REDACTED',
      'appium:noReset': true
    },
    firstMatch: [ {} ]
  }
}
2025-03-14T05:45:11.189Z WARN webdriver: WebDriverError: Request failed with error code ECONNREFUSED when running "http://127.0.0.1:4723/session" with method "POST" and args "{"capabilities":{"alwaysMatch":{"platformName":"Android","appium:automationName":"UiAutomator2","appium:deviceName":"REDACTED","appium:udid":"REDACTED","appium:noReset":true},"firstMatch":[{}]}}"
2025-03-14T05:45:11.189Z INFO webdriver: Retrying 1/3
2025-03-14T05:45:11.189Z INFO webdriver: [POST] http://127.0.0.1:4723/session
2025-03-14T05:45:11.189Z INFO webdriver: DATA {
  capabilities: {
    alwaysMatch: {
      platformName: 'Android',
      'appium:automationName': 'UiAutomator2',
      'appium:deviceName': 'REDACTED',
      'appium:udid': 'REDACTED',
      'appium:noReset': true
    },
    firstMatch: [ {} ]
  }
}
2025-03-14T05:45:11.692Z WARN webdriver: WebDriverError: Request failed with error code ECONNREFUSED when running "http://127.0.0.1:4723/session" with method "POST" and args "{"capabilities":{"alwaysMatch":{"platformName":"Android","appium:automationName":"UiAutomator2","appium:deviceName":"REDACTED","appium:udid":"REDACTED","appium:noReset":true},"firstMatch":[{}]}}"
2025-03-14T05:45:11.692Z INFO webdriver: Retrying 2/3
2025-03-14T05:45:11.693Z INFO webdriver: [POST] http://127.0.0.1:4723/session
2025-03-14T05:45:11.693Z INFO webdriver: DATA {
  capabilities: {
    alwaysMatch: {
      platformName: 'Android',
      'appium:automationName': 'UiAutomator2',
      'appium:deviceName': 'REDACTED',
      'appium:udid': 'REDACTED',
      'appium:noReset': true
    },
    firstMatch: [ {} ]
  }
}
2025-03-14T05:45:12.697Z WARN webdriver: WebDriverError: Request failed with error code ECONNREFUSED when running "http://127.0.0.1:4723/session" with method "POST" and args "{"capabilities":{"alwaysMatch":{"platformName":"Android","appium:automationName":"UiAutomator2","appium:deviceName":"REDACTED","appium:udid":"REDACTED","appium:noReset":true},"firstMatch":[{}]}}"
2025-03-14T05:45:12.698Z INFO webdriver: Retrying 3/3
2025-03-14T05:45:12.698Z INFO webdriver: [POST] http://127.0.0.1:4723/session
2025-03-14T05:45:12.698Z INFO webdriver: DATA {
  capabilities: {
    alwaysMatch: {
      platformName: 'Android',
      'appium:automationName': 'UiAutomator2',
      'appium:deviceName': 'REDACTED',
      'appium:udid': 'REDACTED',
      'appium:noReset': true
    },
    firstMatch: [ {} ]
  }
}
2025-03-14T05:45:12.701Z ERROR webdriver: WebDriverError: Request failed with error code ECONNREFUSED when running "http://127.0.0.1:4723/session" with method "POST" and args "{"capabilities":{"alwaysMatch":{"platformName":"Android","appium:automationName":"UiAutomator2","appium:deviceName":"REDACTED","appium:udid":"REDACTED","appium:noReset":true},"firstMatch":[{}]}}"
2025-03-14T05:45:12.701Z ERROR webdriver: Error: WebDriverError: Request failed with error code ECONNREFUSED when running "http://127.0.0.1:4723/session" with method "POST" and args "{"capabilities":{"alwaysMatch":{"platformName":"Android","appium:automationName":"UiAutomator2","appium:deviceName":"REDACTED","appium:udid":"REDACTED","appium:noReset":true},"firstMatch":[{}]}}"
    at FetchRequest._libRequest (file:///home/jlucaso/projects/temp/bun-3997/node_modules/webdriver/build/node.js:1746:13)
    at process.processTicksAndRejections (node:internal/process/task_queues:105:5)
    at async FetchRequest._request (file:///home/jlucaso/projects/temp/bun-3997/node_modules/webdriver/build/node.js:1756:20)
    at async startWebDriverSession (file:///home/jlucaso/projects/temp/bun-3997/node_modules/webdriver/build/node.js:1093:16)
    at async _WebDriver.newSession (file:///home/jlucaso/projects/temp/bun-3997/node_modules/webdriver/build/node.js:1329:41)
    at async remote (file:///home/jlucaso/projects/temp/bun-3997/node_modules/webdriverio/build/node.js:8722:20)
    at async testAppiumConnection (file:///home/jlucaso/projects/temp/bun-3997/index.ts:14:20)
Error connecting to device: Unable to connect to "http://127.0.0.1:4723/", make sure browser driver is running on that address.
It seems like the service failed to start or is rejecting any connections.
```
